text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+    /// Create a random value of type `Self`
+    fn random<R: ?Sized + RngBase>(data: Data, rng: &mut Rng) -> Self;
+}
+
+// Example:
+
+// create a raw u8 with no restriction.
+impl Random<FullRange> for u8 {
+    ...
+}
+
+// create a u8 between the bounds of range (same exclusivity properties as usual)
+impl Random<Range<u8>> for u8 {
+    ...
+}
+
+
+/// Random number generator.
+trait Rng: io::Reader {
+    /// Generate/fetch the next `u32` (no range restriction)
+    #[unstable]
+    fn next_u32(&mut self) -> u32 { ... }
+
+    /// Generate/fetch the next `u64` (no range restriction)
+    #[unstable]
+    fn next_u64(&mut self) -> u64 { ... }
+    /// Generate/fetch the next `f32` in the range [0,1)
+    #[unstable]
+    fn next_f32(&mut self) -> f32 { ... }
+    /// Generate/fetch the next `f64` in the range [0,1)
+    #[unstable]
+    fn next_f64(&mut self) -> f64 { ... }
+}
+
+impl<'a, T: ?Sized + Rng> Rng for &'a mut T {}
+impl<T: ?Sized + Rng> Rng for Box<T> {}
+
+/// Basic extension functionality to ensure that `Rng` is object-safe.
+trait RngExt {
+    fn gen<T, D>(&mut self, data: D) -> T where T: Random<D> { ... }
+
+    /// Return an iterator that generates an infinite sequence of random values of type `T`.
+    fn gen_iter<T, D>(self, data: D) -> Generator<T, D, Self> where T: Random<D> { ... }
+
+    /// Randomize the order of the elements of `values`
+    fn shuffle<T>(&mut self, values: &mut [T]) { ... }
+    /// Select one value uniformly at random from the slice.
+    #[unstable]
+    fn choose<'a>(&mut self, values: &[T]) -> Option<&'a T> { ... }
+    /// Select `n` values uniformly at random from the iterator.
+    fn sample<I: Iterator>(&mut self, n: uint, values: I) -> Vec<I::Item> { ... }
+}
+
+impl<T: Rng> RngExt for T {}
+```
+
+As today, the relationship between output of the methods of `Rng`, and
+between them and `io::Reader` is unspecified.
+
+### The `Random` trait
+
+The change to `Random` (currently `Rand`) is motivated by allowing
+`gen` and `gen_range` to be unified, what was previously
+`rng.gen_range(a, b)` can become `rng.gen(a..b)`. However, this makes
+the non-ranged generation slightly more complicated/"magic
+incantation"-y: `rng.gen(..)` (this relies on having explicit syntax
+for `FullRange`, per [RFC 702]).
+
+[RFC 702]: https://github.com/rust-lang/rfcs/pull/702
+
+The convention for using `FullRange` for unrestricted generation may
+seem peculiar for types that don't have any concept of ranges,
+e.g. creating a randomly seeded RNG, but the consistency and
+"vagueness" of `..` likely means this is OK.
+
+(NB. general discussion, non-normative) This approach allows us to
+naturally leverage any "range combinators" used elsewhere,
+e.g. inclusive endpoints, and so remove the need for the `Closed01`
+and `Open01` types. It also allows us to implement `Random` for
+e.g. `Vec`, say, `rng.gen::<Vec<u8>>(10..20)` would create a vector of
+some length between 10 and 20 and fill it with `u8`s.
```

Comment:
For 1.0 I really suggest avoiding such implementations. If coherence rules are relaxed sometime later, this becomes a perfect “users can implement this themselves” material. If not… well, adding something like this could be considered then.